llvm/include/llvm/Frontend/Offloading/Utility.h

```diff
@@ -159,6 +160,107 @@ namespace intel {
 /// the Intel runtime offload plugin.
 Error containerizeOpenMPSPIRVImage(std::unique_ptr<MemoryBuffer> &Binary);
 } // namespace intel
+
+namespace sycl {
+class PropertySetRegistry;
+
+// A property value. It can be either a 32-bit unsigned integer or a byte array.
+class PropertyValue {
+public:
+  using ByteArrayTy = SmallVector<char, 8>;
+
+  PropertyValue() = default;
+  PropertyValue(uint32_t Val) : Value(Val) {}
+  PropertyValue(StringRef Data)
+      : Value(ByteArrayTy(Data.begin(), Data.end())) {}
+
+  template <typename C, typename T = typename C::value_type>
+  PropertyValue(const C &Data)
+      : PropertyValue({reinterpret_cast<const char *>(Data.data()),
+                       Data.size() * sizeof(T)}) {}
+
+  uint32_t asUint32() const {
+    assert(getType() == PV_UInt32 && "must be UINT32 value");
+    return std::get<uint32_t>(Value);
+  }
+
+  StringRef asByteArray() const {
+    assert(getType() == PV_ByteArray && "must be BYTE_ARRAY value");
+    const auto &ByteArrayRef = std::get<ByteArrayTy>(Value);
+    return {ByteArrayRef.data(), ByteArrayRef.size()};
+  }
+
+  // Note: each enumeration value corresponds one-to-one to the
+  // index of the std::variant.
+  enum Type { PV_None = 0, PV_UInt32 = 1, PV_ByteArray = 2 };
+  Type getType() const { return static_cast<Type>(Value.index()); }
+
+  bool operator==(const PropertyValue &Rhs) const { return Value == Rhs.Value; }
+
+private:
+  std::variant<std::monostate, std::uint32_t, ByteArrayTy> Value = {};
+};
+
+/// A property set is a collection of PropertyValues, each identified by a name.
+/// A property set registry is a collection of property sets, each identified by
+/// a category name. The registry allows for the addition, removal, and
+/// retrieval of property sets and their properties.
+class PropertySetRegistry {
+  using PropertyMapTy = StringMap<unsigned>;
+  /// A property set. Preserves insertion order when iterating elements.
+  using PropertySet = MapVector<SmallString<16>, PropertyValue, PropertyMapTy>;
+  using MapTy = MapVector<SmallString<16>, PropertySet, PropertyMapTy>;
```

Comment:
I would discourage usage of `SmallString<16>`.
https://github.com/intel/llvm/blob/sycl/sycl/doc/design/PropertySets.md
Typically, keys are larger than 16 and we wouldn't make use of this limit.
Also, LLVM's Programmers Manual discourages putting SmallString `as the member of a frequently-allocated heap data structure`.

I would rather make it `std::map<std::string, std::map<std::string, PropertyValue>>`.

We aren't going to have many keys here and it would make the interface extremely simple.